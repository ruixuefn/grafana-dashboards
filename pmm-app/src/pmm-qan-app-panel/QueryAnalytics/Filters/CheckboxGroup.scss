.filter-label {
  display: grid;
  grid-template-areas: 'filtername percentagearea';
  grid-template-rows: 30px;
  grid-template-columns: 160px auto;
  grid-gap: 10px;
  height: auto;
  margin: 0;
}

.filter-name {
  grid-area: filtername;
}

.percentage {
  grid-area: percentagearea;
  display: flex;
  justify-content: flex-end;
}
