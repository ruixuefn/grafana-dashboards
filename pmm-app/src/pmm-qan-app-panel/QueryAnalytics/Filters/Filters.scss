.query-analytics-filters-wrapper {
  border: 1px solid rgb(40, 40, 40);
  padding: 5px 10px;
  overflow-y: scroll;
}

.filters-header {
  display: flex;
  align-items: flex-end;
}
