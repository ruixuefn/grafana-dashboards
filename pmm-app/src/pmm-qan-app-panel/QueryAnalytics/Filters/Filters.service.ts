import { apiRequest } from '../../../react-plugins-deps/components/helpers/api';

class FiltersService {
  static getAutocompleteFiltersList(selectedVariables) {
    const selected: any[] = [];
    const options: any[] = [];
    const labels = result.labels;

    Object.keys(labels).forEach(label => {
      labels[label].name.forEach(metric => {
        const passedVariables = selectedVariables[label];

        metric.checked = passedVariables && passedVariables.some(variable => variable === metric.value);

        metric.key = `${label}:${metric.value}`;
        // @ts-ignore
        options.push(metric);
        if (metric.checked) {
          // @ts-ignore
          selected.push(`${label}:${metric.value}`);
        }
      });
    });

    return { selected, options };
  }

  static async getQueryOverviewFiltersList(paramLabels, from, to) {
    const { labels } = await apiRequest.post<any, any>('/v0/qan/Filters/Get', {
      labels: Object.keys(paramLabels).map(key => {
        return {
          key: key,
          value: paramLabels[key],
        };
      }),
      main_metric_name: '',
      period_start_from: from,
      period_start_to: to,
    });

    Object.keys(labels).forEach(label => {
      labels[label].name.forEach(metric => {
        const passedVariables = paramLabels[label];
        metric.checked = passedVariables && passedVariables.some(variable => variable === metric.value);
      });
    });

    return labels;
  }
}

export default FiltersService;
