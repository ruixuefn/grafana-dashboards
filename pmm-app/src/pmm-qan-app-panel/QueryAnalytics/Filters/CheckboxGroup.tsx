import React, { useState } from 'react';
import { CheckboxField } from '../../../react-plugins-deps/components/FormComponents/Checkbox/Checkbox';
import { Humanize } from '../../../react-plugins-deps/components/helpers/Humanization';
import { Divider } from 'antd';
import './CheckboxGroup.scss';

const TOP_LIMIT = 5;

export const CheckboxGroup = ({ form, name, items, group, showAll, filter, labels }) => {
  const [showTop, setShowTop] = useState(false);
  const data = showTop ? items.slice(0, TOP_LIMIT) : items;
  const itemsList = data
    .filter(item => item.value)
    .filter(item => {
      if (!showAll) {
        return item.checked;
      }
      return true;
    })
    .filter(item => item.value.toLowerCase().includes(filter.toLowerCase()))
    .map(item => {
      return (
        <div className={'filter-label'}>
          <span className={'filter-name'}>
            <CheckboxField
              form={form}
              name={`${group}:${item.value}`}
              label={item.value}
              checked={labels && labels[group] && labels[group].includes(item.value)}
            />
          </span>
          <span className={'percentage'}>
            <span>{Humanize.transform(item.main_metric_percent, 'percent')}</span>
          </span>
        </div>
      );
    });
  return itemsList.length ? (
    <div>
      <p style={{ display: 'flex', justifyItems: 'space-between', marginBottom: '0', marginTop: '20px' }}>
        <span style={{ marginRight: 'auto' }}>{name}</span>
        {items.length > TOP_LIMIT ? (
          <span
            onClick={() => {
              setShowTop(!showTop);
            }}
            style={{ cursor: 'pointer' }}
          >
            {showTop ? `Show all (${items.length})` : `Show top ${TOP_LIMIT}`}
          </span>
        ) : (
          <span></span>
        )}
      </p>
      <Divider style={{ marginTop: '0', marginBottom: '5px' }}></Divider>
      {itemsList}
    </div>
  ) : null;
};
