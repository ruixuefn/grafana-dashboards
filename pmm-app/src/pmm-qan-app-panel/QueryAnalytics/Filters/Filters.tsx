import React, { ReactElement, useContext, useEffect, useState } from 'react';
import { Humanize } from '../../../react-plugins-deps/components/helpers/Humanize';
import './Filters.scss';
import { Divider } from 'antd';
import { ManagementContext, StateContext } from '../StateContext';
import FiltersService from './Filters.service';
import { CheckboxField } from '../../../react-plugins-deps/components/FormComponents/Checkbox/Checkbox';
import { useForm } from 'react-final-form-hooks';
import { Form as FormFinal } from 'react-final-form';
import Search from 'antd/es/input/Search';
import set = Reflect.set;
const humanize = new Humanize();

const checkboxGroup = (form, name, items, group, showAll, filter) => {
  const itemsList = items
    .filter(item => item.value)
    .filter(item => {
      if (!showAll) {
        return item.checked;
      }
      return true;
    })
    .filter(item => item.value.toLowerCase().includes(filter.toLowerCase()))
    .map(item => {
      return (
        <div className={'filter-label'}>
          <span className={'filter-name'}>
            <CheckboxField form={form} name={`${group}:${item.value}`} label={item.value} checked={item.checked} />
          </span>
          <span className={'percentage'}>
            <span>{humanize.transform(item.main_metric_percent, 'percent')}</span>
          </span>
        </div>
      );
    });
  return itemsList.length ? (
    <div>
      <Divider>{name}</Divider>
      {itemsList}
    </div>
  ) : null;
};

const checkboxGroups = [
  {
    name: 'Environment',
    dataKey: 'environment',
  },
  {
    name: 'Cluster',
    dataKey: 'cluster',
  },
  {
    name: 'Replication Set',
    dataKey: 'replication_set',
  },
  {
    name: 'Database',
    dataKey: 'database',
  },
  {
    name: 'Schema',
    dataKey: 'schema',
  },
  {
    name: 'Node Name',
    dataKey: 'node_name',
  },
  {
    name: 'Service Name',
    dataKey: 'service_name',
  },
  {
    name: 'Client Host',
    dataKey: 'client_host',
  },
  {
    name: 'User Name',
    dataKey: 'username',
  },
  {
    name: 'Service Type',
    dataKey: 'service_type',
  },
  {
    name: 'Node Type',
    dataKey: 'node_type',
  },
  {
    name: 'City',
    dataKey: 'city',
  },
  {
    name: 'Availability Zone',
    dataKey: 'az',
  },
];

const Filters = () => {
  const [showAll, showSetAll] = useState(true);
  const [filter, setFilter] = useState('');
  const context = useContext(StateContext);
  console.log(context);
  const filters = FiltersService.getQueryOverviewFiltersList(context.state.labels || {});

  return (
    <FormFinal
      onSubmit={() => {}}
      render={(): ReactElement => {
        const { form, handleSubmit } = useForm({
          onSubmit: filters => {
            console.log('submitting');
          },
          validate: filters => {
            // TODO: temp solution, need to figure out why handleSubmit works wrong
            console.log('set labels called');
            // context.setLabels(filters);
            // context.dispatch({
            //   type: 'SET_LABELS',
            //   payload: filters,
            // });
          },
          initialValues: {},
        });
        // @ts-ignore
        return (
          <form onSubmit={handleSubmit} className="add-instance-form app-theme-dark">
            <div className={'filters-header'} style={{ padding: '5px 0px', height: '50px' }}>
              <h5 style={{ margin: '3px', marginRight: '15px' }}>Filters</h5>
              {showAll ? (
                <a href="#" className={'filter-switchers'} onClick={showSetAll.bind(null, false)}>
                  Show Selected
                </a>
              ) : (
                <a href="#" className={'filter-switchers'} onClick={showSetAll.bind(null, true)}>
                  Show All
                </a>
              )}
              <a
                href="#"
                className={'filter-switchers'}
                style={{ marginLeft: 'auto' }}
                onClick={() => {
                  context.dispatch({ type: 'RESET_LABELS' });
                }}
              >
                Reset All
              </a>
            </div>
            <div className={'query-analytics-filters-wrapper'}>
              <Search placeholder="Filters search..." onChange={e => setFilter(e.target.value)} style={{ width: '100%' }} />
              {checkboxGroups.map(group => {
                return checkboxGroup(form, group.name, filters[group.dataKey].name, group.dataKey, showAll, filter);
              })}
            </div>
          </form>
        );
      }}
    />
  );
};

export default Filters;
