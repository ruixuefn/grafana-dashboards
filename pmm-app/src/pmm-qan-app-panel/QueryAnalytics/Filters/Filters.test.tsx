// Just a stub test
import { Filters, FILTERS_GROUPS } from './Filters';
import React from 'react';
import renderer from 'react-test-renderer';
import { useForm } from 'react-final-form-hooks';
import { renderHook } from '@testing-library/react-hooks';

jest.mock('antd/es/input/Search', () => () => {
  return <></>;
});

jest.mock('react-scrollbar', () => () => {
  return <div></div>;
});

jest.mock('../../../react-plugins-deps/components/helpers/notification-manager', () => () => ({}));

const MOCK_FILTERS = {
  az: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  client_host: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  cluster: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  container_id: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  container_name: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  database: {
    name: [
      {
        value: 'postgres',
        main_metric_percent: 0.99822205,
        main_metric_per_sec: 0.0028915745,
        checked: false,
      },
      {
        value: 'pmm-managed',
        main_metric_percent: 0.001778005,
        main_metric_per_sec: 0.000005150391,
        checked: true,
      },
    ],
  },
  environment: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  machine_id: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  node_id: {
    name: [
      {
        value: 'pmm-server',
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  node_model: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  node_name: {
    name: [
      {
        value: 'pmm-server',
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  node_type: {
    name: [
      {
        value: 'generic',
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  region: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  replication_set: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  schema: {
    name: [
      {
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  service_id: {
    name: [
      {
        value: '/service_id/1b879be7-cf43-487a-82f0-09143f8d914b',
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  service_name: {
    name: [
      {
        value: 'pmm-server-postgresql',
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  service_type: {
    name: [
      {
        value: 'postgresql',
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
  username: {
    name: [
      {
        value: 'pmm-managed',
        main_metric_percent: 1,
        main_metric_per_sec: 0.000005150391,
      },
    ],
  },
};
describe('Filters test', () => {
  it('Renders correct with right props', () => {
    const { result } = renderHook(() => {
      return useForm({
        onSubmit: values => {
          console.log(values);
        },
      });
    });

    const component = renderer.create(
      <Filters groups={FILTERS_GROUPS} filters={MOCK_FILTERS} labels={{}} dispatch={() => {}} form={result.current.form} />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
