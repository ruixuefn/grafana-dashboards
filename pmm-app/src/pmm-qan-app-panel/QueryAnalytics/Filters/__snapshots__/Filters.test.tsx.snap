// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters test Renders correct with right props 1`] = `
<div>
  <div
    className="css-1mxzbbx"
  >
    <h5
      className="css-1urorti"
    >
      Filters
    </h5>
    <button
      className="ant-btn css-x930nf ant-btn-link"
      onClick={[Function]}
      type="button"
    >
      <span>
        Show Selected
      </span>
    </button>
    <button
      className="ant-btn css-vuuytn ant-btn-link"
      onClick={[Function]}
      type="button"
    >
      <span>
        Reset All
      </span>
    </button>
  </div>
  <div />
</div>
`;
